hypershift-operator/controllers/nodepool/nodepool_controller.go

```diff
@@ -2917,3 +2918,140 @@ func aggregateMachineMessages(msgs []string) string {
 
 	return builder.String()
 }
+
+// secretJanitor reconciles secrets and determines which secrets should remain in the cluster and which should be cleaned up.
+// Any secret annotated with a nodePool name should only be on the cluster if the nodePool continues to exist
+// and if our current calculation for the inputs to the name matches what the secret is named.
+type secretJanitor struct {
+	*NodePoolReconciler
+}
+
+func (r *secretJanitor) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	secret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, req.NamespacedName, secret); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("not found", "request", req.String())
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "error getting secret")
+		return ctrl.Result{}, err
+	}
+
+	nodePoolName, annotated := secret.Annotations[nodePoolAnnotation]
+	if !annotated {
+		return ctrl.Result{}, nil
```

Comment:
I'd expect this reconciler to return early here as well when
`if !isTokenSecret && !isUserdataSecret`